processing/src/main/java/org/apache/druid/segment/CursorBuildSpec.java

```diff
@@ -0,0 +1,264 @@
+    this.virtualColumns = virtualColumns;
+    this.aggregators = aggregators;
+    this.orderByColumns = preferredOrdering;
+    this.queryContext = queryContext;
+    this.queryMetrics = queryMetrics;
+  }
+
+  @Nullable
+  public Filter getFilter()
+  {
+    return filter;
+  }
+
+  public Interval getInterval()
+  {
+    return interval;
+  }
+
+  @Nullable
+  public List<String> getGroupingColumns()
+  {
+    return groupingColumns;
+  }
+
+  public VirtualColumns getVirtualColumns()
+  {
+    return virtualColumns;
+  }
+
+  @Nullable
+  public List<AggregatorFactory> getAggregators()
+  {
+    return aggregators;
+  }
+
+  @Nullable
+  public List<OrderBy> getPreferredOrdering()
+  {
+    return orderByColumns;
+  }
+
+  public QueryContext getQueryContext()
+  {
+    return queryContext;
+  }
+
+  @Nullable
+  public QueryMetrics<?> getQueryMetrics()
+  {
+    return queryMetrics;
+  }
+
+  public static boolean preferDescendingTimeOrder(@Nullable List<OrderBy> preferredOrdering)
+  {
+    if (preferredOrdering != null && preferredOrdering.size() == 1) {
+      final OrderBy orderBy = Iterables.getOnlyElement(preferredOrdering);
+      return ColumnHolder.TIME_COLUMN_NAME.equals(orderBy.getColumnName()) && Order.DESCENDING == orderBy.getOrder();
+    }
+    return false;
+  }
+
+  public static class CursorBuildSpecBuilder
+  {
+    @Nullable
+    private Filter filter;
+    private Interval interval = Intervals.ETERNITY;
+
+    @Nullable
+    private List<String> groupingColumns;
+    private VirtualColumns virtualColumns = VirtualColumns.EMPTY;
+    @Nullable
+    private List<AggregatorFactory> aggregators;
+    @Nullable
+    private List<OrderBy> preferredOrdering;
+
+    private QueryContext queryContext = QueryContext.empty();
+    @Nullable
+    private QueryMetrics<?> queryMetrics;
+
+    private CursorBuildSpecBuilder()
+    {
+      // initialize with defaults
+    }
+
+    private CursorBuildSpecBuilder(CursorBuildSpec buildSpec)
+    {
+      this.filter = buildSpec.filter;
+      this.interval = buildSpec.interval;
+      this.groupingColumns = buildSpec.groupingColumns;
+      this.virtualColumns = buildSpec.virtualColumns;
+      this.aggregators = buildSpec.aggregators;
+      this.preferredOrdering = buildSpec.orderByColumns;
+      this.queryContext = buildSpec.queryContext;
+      this.queryMetrics = buildSpec.queryMetrics;
+    }
+
+    public CursorBuildSpecBuilder setFilter(@Nullable Filter filter)
+    {
+      this.filter = filter;
+      return this;
+    }
+
+    public CursorBuildSpecBuilder setInterval(Interval interval)
+    {
+      this.interval = interval;
+      return this;
+    }
+
+    public CursorBuildSpecBuilder setGroupingAndVirtualColumns(
+        Granularity granularity,
+        @Nullable List<String> groupingColumns,
+        VirtualColumns virtualColumns
+    )
+    {
+      final VirtualColumn granularityVirtual = Granularities.toVirtualColumn(granularity);
```

Comment:
yea, i guess doing the query -> buildspec translation in the engines would make more sense which i think also would satisfy many of your complaints, will move it there.

as for the other thing, maybe we should make a specialized time floor virtual column to use instead of expression version, will think about it